src/main/java/htsjdk/variant/variantcontext/writer/BCF2Writer.java

```diff
@@ -221,6 +221,35 @@ public void close() {
         super.close();
     }
 
+    @Override
+    public void setVcfHeader(VCFHeader header) {
+        header = new VCFHeader(header.getMetaDataInSortedOrder(), header.getGenotypeSamples());
+        this.header = doNotWriteGenotypes ? new VCFHeader(header.getMetaDataInSortedOrder()) : header;
+        // create the config offsets map
+        if ( header.getContigLines().isEmpty() ) {
+            if ( ALLOW_MISSING_CONTIG_LINES ) {
+                if ( GeneralUtils.DEBUG_MODE_ENABLED ) {
+                    System.err.println("No contig dictionary found in header, falling back to reference sequence dictionary");
+                }
+                createContigDictionary(VCFUtils.makeContigHeaderLines(getRefDict(), null));
+            } else {
+                throw new IllegalStateException("Cannot write BCF2 file with missing contig lines");
+            }
+        } else {
+            createContigDictionary(header.getContigLines());
+        }
```

Comment:
Much of this code is duplicated in writeHeader, and should be factored out into a common method called by both.